cabal-install/Distribution/Client/Dependency.hs

```diff
@@ -587,13 +587,19 @@ interpretPackagesPreference :: Set PackageName
                             -> (PackageName -> PackagePreferences)
 interpretPackagesPreference selected defaultPref prefs =
   \pkgname -> PackagePreferences (versionPref pkgname) (installPref pkgname)
-
   where
     versionPref pkgname =
-      fromMaybe anyVersion (Map.lookup pkgname versionPrefs)
-    versionPrefs = Map.fromList
-      [ (pkgname, pref)
-      | PackageVersionPreference pkgname pref <- prefs ]
+      fromMaybe [anyVersion] (Map.lookup pkgname versionPrefs)
+    versionPrefs = foldr mergeIfMultiplePrefs Map.empty prefs
+    mergeIfMultiplePrefs (PackageVersionPreference pkgname pref) m =
+      Map.insertWith (++) pkgname [pref] m
+    -- Previously the two cases below were simply ignored, and I think for a good
```

Comment:
See the comment here - how do we handle "installed" or "latest" preference in the input pref list? Previously they were simply ignored by a partial list comprehension.